### PR TITLE
Fix MM53643

### PR DESCRIPTION
**Pull request by larkox**
#### Summary
Fix MM53643

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-53643

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Comment by agnivade** [comment]
Have you tested this well in an HA env? This code is extremely tricky and fails in a myriad of ways in HA.

**Comment by larkox** [comment]
@agnivade I have not, but not sure how HA may affect this in a way we can handle.

If the DB (or its own read replica) says the post exist, the post exist. If the DB says the post doesn't exist, it will clear any metadata "just in case", before sending the websocket event.

**Comment by agnivade** [comment]
Yes it's unlikely at first glance, but we won't be sure unless we actually test it. There are possibly other unrelated parts which might fail in HA. Just like it did in my original PR.

**Review by jupenur — APPROVED** [review]

**Comment by larkox** [comment]
@agnivade On the test server, following the reproduction steps, I see weird behavior that does not seem related to the PR. Mainly, the kind of behavior I am seeing is that not all the websocket events are being sent properly (some of them are never sent). That creates a more erratic behavior, but I would say it is expected.

0/5 if this is something we should look more into, but does seem unrelated to this PR.

**Review by agnivade — COMMENTED** [review]
Wondering if we can add a unit test for this?

We can add a test in the api layer where we create a post with metadata and then verify that the websocket event is correctly generated. There's a sample test in `TestDeletePostEvent` in api4/post_test.go.

**Comment by agnivade** [comment]
> That creates a more erratic behavior, but I would say it is expected.

If some websocket events are not sent properly, I wouldn't say it's expected. I'd file a ticket with any repro steps if possible and track it atleast. There might be an unrelated bug somewhere.

**Comment by larkox** [comment]
@agnivade
Re: Unit Test
Didn't know we can test the websocket events like that! Will add those unit tests.

Re: Erratic behavior
What I meant is that if the events are not reaching, it is normal it is erratic :P But sure, the events not reaching seems like a bug. I will double check (to make sure it is not something that "comes and goes") and create a ticket.

**Comment by hmhealey** [comment]
@larkox I think you hit that weird situation I hit before with the "HA server" label where one server in the cluster gets a license and the other does not. Since the cluster is messed up, WS events don't always send successfully. Since you pushed a new commit, that restarted the cluster so it seems to be running fine now

**Review by hmhealey — APPROVED** [review]

File: server/channels/app/notification_test.go [hunk 1]
```diff
@@ -395,6 +395,51 @@ func TestSendNotifications_MentionsFollowers(t *testing.T) {
...
+		assert.Equal(t, postURL, received1Post.Message)
+		assert.Nil(t, received1Post.Metadata.Embeds)
+
+		received2 := <-messages2
+		require.Equal(t, model.WebsocketEventPosted, received2.EventType())
+		received2Post := &model.Post{}
+		err = json.Unmarshal([]byte(received2.GetData()["post"].(string)), &received2Post)
+		require.NoError(t, err)
+		assert.Equal(t, postURL, received2Post.Message)
+		assert.Nil(t, received1Post.Metadata.Embeds)
```
Comment (agnivade) [inline server/channels/app/notification_test.go:441]: Why are we creating 2 connections? Both seem to be testing for the exact same thing.

Comment (larkox) [reply server/channels/app/notification_test.go:441]: You are right. Originally I wanted to test different things for each user, but did not make sense and I forgot to revert this part. Will change.

**Review by agnivade — APPROVED** [review]

**Pull request merged**